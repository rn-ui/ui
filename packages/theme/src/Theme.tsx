/* tslint:disable */
import deepMerge from './tools/deepMerge';
import mergeComponentAndThemes from './tools/mergeComponentAndThemes';
import { resolveIncludes } from './tools/resolveIncludes';
import normalizeTheme from './normalizer/normalizeTheme';
import defaultTheme from './resources/default';

interface Themes {
  default: object;
  light: object;
  dark: object;
}

// @ts-ignore
const themes: Themes = {};

// Privates, ideally those should be symbols
const THEME_STYLES = '@southem.theme';

const resolveStyle = (style, baseStyle) => normalizeTheme(resolveIncludes(style, baseStyle));

/**
 * Theme
 *
 * @example
 * const defaultTheme = {};
 * const defaultThemePro = {};
 *
 * Theme.registerTheme('default', [
 *   defaultTheme,
 *   defaultThemePro,
 * ]);
 */
class Theme {

  constructor(themeStyle) {
    // @ts-ignore
    // this[THEME_STYLES] = resolveStyle(themeStyle);
  }

  /**
   * registerTheme
   * @param {String} themeName
   * @param {Array|Object} styleSheets
   */
  static registerTheme(themeName, styleSheets: Array<Object>|Object): void {
    let mergedStyleSheets = styleSheets;

    if (Array.isArray(styleSheets)) {
      mergedStyleSheets = deepMerge({}, ...styleSheets);
    }

    // themes[themeName] = new Theme(themeName, mergedStyleSheets);
    // @ts-ignore
    themes[themeName] = resolveStyle(mergedStyleSheets);
  }

  /**
   * registerDefaultTheme
   * @param {Object} styleSheets
   */
  static registerDefaultTheme(styleSheets): void {
    Theme.registerTheme('default', styleSheets);
  }

  /**
   * getTheme
   * @param {String} themeName
   * @return {Object}
   */
  static getTheme(themeName): Object {
    // @ts-ignore
    if (themeName === 'default' && !themes.default) {
      Theme.registerDefaultTheme(defaultTheme);
    }

    return themes[themeName];
  }

  /**
   * getDefaultTheme
   * @return {Object}
   */
  static getDefaultTheme(): Object {
    return Theme.getTheme('default');
  }

  /**
   * isDefine
   * @param {String} themeName
   * @return {Boolean}
   */
  static isDefine(themeName): Boolean {
    return !!themes[themeName];
  }
}

export {
  Theme as default,
  themes,
};
