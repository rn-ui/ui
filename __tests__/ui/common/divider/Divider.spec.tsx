import React from 'react';
import {
  View,
  StyleSheet,
} from 'react-native';
import {
  render,
  shallow,
  RenderAPI,
} from 'react-native-testing-library';
import { ReactTestInstance } from 'react-test-renderer';
import Theme, { ThemeProvider } from '@southem/theme';
import {
  darkTheme,
  Divider,
  DividerProps,
} from '@southem/ui';

Theme.registerDefaultTheme(darkTheme);

const Mock = (props?: DividerProps): React.ReactElement<{}> => {
  return (
    <ThemeProvider
      theme={'default'}>
      <Divider {...props} />
    </ThemeProvider>
  );
};

const renderComponent = (props?: DividerProps): RenderAPI => {
  return render(
    <Mock {...props} />,
  );
};

describe('@divider: matches snapshot', () => {
  describe('* interaction', () => {
    it('* stateless', () => {
      const component: RenderAPI = renderComponent();
      const { output } = shallow(component.getByType(Divider));

      expect(output).toMatchSnapshot();
    });
  });
  describe('* appearance', () => {

    const text: React.ReactText = 'TEXT';

    it('* empty', () => {
      const component: RenderAPI = renderComponent();
      const { output } = shallow(component.getByType(Divider));

      expect(output).toMatchSnapshot();
    });

    it('* text', () => {
      const component: RenderAPI = renderComponent({ children: text });
      const { output } = shallow(component.getByType(Divider));

      expect(output).toMatchSnapshot();
    });

    it('* text * align items center and justify content of center', () => {
      const component: RenderAPI = renderComponent({
        children: text,
      });
      const componentDivider: ReactTestInstance = component.getByType(View);

      const styles  = StyleSheet.flatten(componentDivider.props.style);
      const { output } = shallow(componentDivider);

      expect(styles.alignItems).toEqual('center');
      expect(styles.justifyContent).toBe('center');
      expect(output).toMatchSnapshot();
    });

    it('* custom text (styled)', () => {
      const component: RenderAPI = renderComponent({
        children: text,
        // @ts-ignore
        style: [{ color: 'red' }, { fontSize: 30 }],
      });
      const componentDivider: ReactTestInstance = component.getByType(View);

      const styles  = StyleSheet.flatten(componentDivider.props.style);
      const { output } = shallow(componentDivider);

      expect(styles).toEqual({
        alignItems: 'center',
        color: 'red',
        fontSize: 30,
        justifyContent: 'center',
      });

      expect(output).toMatchSnapshot();
    });
  });

});
