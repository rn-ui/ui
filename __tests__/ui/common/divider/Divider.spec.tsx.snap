// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@divider: matches snapshot * appearance * custom text (styled) 1`] = `
<View
  animationOptions={Object {}}
  onLayout={[Function]}
  orientation="horizontal"
  style={
    Object {
      "alignItems": "center",
      "color": "red",
      "fontSize": 30,
      "justifyContent": "center",
    }
  }
>
  <ForwardRef(WrappingElement)
    style={null}
  />
  <ForwardRef(WrappingElement)
    style={null}
  >
    TEXT
  </ForwardRef(WrappingElement)>
  <ForwardRef(WrappingElement)
    style={null}
  />
</View>
`;

exports[`@divider: matches snapshot * appearance * empty 1`] = `
<@theme(Animated(Divider))
  animationOptions={Object {}}
  forwardedRef={null}
/>
`;

exports[`@divider: matches snapshot * appearance * text * align items center and justify content of 50 1`] = `
<View
  animationOptions={Object {}}
  onLayout={[Function]}
  orientation="horizontal"
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <ForwardRef(WrappingElement)
    style={null}
  />
  <ForwardRef(WrappingElement)
    style={null}
  >
    TEXT
  </ForwardRef(WrappingElement)>
  <ForwardRef(WrappingElement)
    style={null}
  />
</View>
`;

exports[`@divider: matches snapshot * appearance * text * align items center and justify content of center 1`] = `
<View
  animationOptions={Object {}}
  onLayout={[Function]}
  orientation="horizontal"
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <ForwardRef(WrappingElement)
    style={null}
  />
  <ForwardRef(WrappingElement)
    style={null}
  >
    TEXT
  </ForwardRef(WrappingElement)>
  <ForwardRef(WrappingElement)
    style={null}
  />
</View>
`;

exports[`@divider: matches snapshot * appearance * text 1`] = `
<@theme(Animated(Divider))
  animationOptions={Object {}}
  forwardedRef={null}
>
  TEXT
</@theme(Animated(Divider))>
`;

exports[`@divider: matches snapshot * interaction * stateless 1`] = `
<@theme(Animated(Divider))
  animationOptions={Object {}}
  forwardedRef={null}
/>
`;
